// warthog/dist/test/modules/dish/dish.resolver.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DishResolver = exports.DishConnectionWhereArgs = exports.ConnectionPageInputOptions = exports.DishConnection = exports.DishEdge = void 0;
var type_graphql_1 = require("type-graphql");
var typedi_1 = require("typedi");
var class_validator_1 = require("class-validator");
var __1 = require("../../../");
var generated_1 = require("../../generated");
var kitchen_sink_model_1 = require("../kitchen-sink/kitchen-sink.model");
var dish_model_1 = require("./dish.model");
var dish_service_1 = require("./dish.service");
var DishEdge = /** @class */ (function () {
    function DishEdge() {
    }
    __decorate([
        type_graphql_1.Field(function () { return dish_model_1.Dish; }, { nullable: false }),
        __metadata("design:type", dish_model_1.Dish)
    ], DishEdge.prototype, "node", void 0);
    __decorate([
        type_graphql_1.Field(function () { return String; }, { nullable: false }),
        __metadata("design:type", String)
    ], DishEdge.prototype, "cursor", void 0);
    DishEdge = __decorate([
        type_graphql_1.ObjectType()
    ], DishEdge);
    return DishEdge;
}());
exports.DishEdge = DishEdge;
var DishConnection = /** @class */ (function () {
    function DishConnection() {
    }
    __decorate([
        type_graphql_1.Field(function () { return type_graphql_1.Int; }, { nullable: false }),
        __metadata("design:type", Number)
    ], DishConnection.prototype, "totalCount", void 0);
    __decorate([
        type_graphql_1.Field(function () { return [DishEdge]; }, { nullable: false }),
        __metadata("design:type", Array)
    ], DishConnection.prototype, "edges", void 0);
    __decorate([
        type_graphql_1.Field(function () { return __1.PageInfo; }, { nullable: false }),
        __metadata("design:type", __1.PageInfo)
    ], DishConnection.prototype, "pageInfo", void 0);
    DishConnection = __decorate([
        type_graphql_1.ObjectType()
    ], DishConnection);
    return DishConnection;
}());
exports.DishConnection = DishConnection;
var ConnectionPageInputOptions = /** @class */ (function () {
    function ConnectionPageInputOptions() {
    }
    __decorate([
        type_graphql_1.Field(function () { return type_graphql_1.Int; }, { nullable: true }),
        class_validator_1.Min(0),
        __metadata("design:type", Number)
    ], ConnectionPageInputOptions.prototype, "first", void 0);
    __decorate([
        type_graphql_1.Field(function () { return String; }, { nullable: true }),
        __metadata("design:type", String)
    ], ConnectionPageInputOptions.prototype, "after", void 0);
    __decorate([
        type_graphql_1.Field(function () { return type_graphql_1.Int; }, { nullable: true }),
        class_validator_1.Min(0),
        __metadata("design:type", Number)
    ], ConnectionPageInputOptions.prototype, "last", void 0);
    __decorate([
        type_graphql_1.Field(function () { return String; }, { nullable: true }),
        __metadata("design:type", String)
    ], ConnectionPageInputOptions.prototype, "before", void 0);
    ConnectionPageInputOptions = __decorate([
        type_graphql_1.ArgsType()
    ], ConnectionPageInputOptions);
    return ConnectionPageInputOptions;
}());
exports.ConnectionPageInputOptions = ConnectionPageInputOptions;
var DishConnectionWhereArgs = /** @class */ (function (_super) {
    __extends(DishConnectionWhereArgs, _super);
    function DishConnectionWhereArgs() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    __decorate([
        type_graphql_1.Field(function () { return generated_1.DishWhereInput; }, { nullable: true }),
        __metadata("design:type", generated_1.DishWhereInput)
    ], DishConnectionWhereArgs.prototype, "where", void 0);
    __decorate([
        type_graphql_1.Field(function () { return generated_1.DishOrderByEnum; }, { nullable: true }),
        __metadata("design:type", String)
    ], DishConnectionWhereArgs.prototype, "orderBy", void 0);
    DishConnectionWhereArgs = __decorate([
        type_graphql_1.ArgsType()
    ], DishConnectionWhereArgs);
    return DishConnectionWhereArgs;
}(ConnectionPageInputOptions));
exports.DishConnectionWhereArgs = DishConnectionWhereArgs;
var DishResolver = /** @class */ (function () {
    function DishResolver(service) {
        this.service = service;
    }
    DishResolver.prototype.kitchenSink = function (dish, ctx) {
        return ctx.dataLoader.loaders.Dish.kitchenSink.load(dish);
    };
    DishResolver.prototype.dishes = function (_a, fields) {
        var where = _a.where, orderBy = _a.orderBy, limit = _a.limit, offset = _a.offset;
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_b) {
                return [2 /*return*/, this.service.find(where, orderBy, limit, offset, fields)];
            });
        });
    };
    DishResolver.prototype.dishConnection = function (_a, fields) {
        var where = _a.where, orderBy = _a.orderBy, pageOptions = __rest(_a, ["where", "orderBy"]);
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_b) {
                return [2 /*return*/, this.service.findConnection(where, orderBy, pageOptions, fields)];
            });
        });
    };
    DishResolver.prototype.dish = function (where) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.service.findOne(where)];
            });
        });
    };
    DishResolver.prototype.createDish = function (data, userId) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.service.create(data, userId)];
            });
        });
    };
    DishResolver.prototype.updateDish = function (_a, userId) {
        var data = _a.data, where = _a.where;
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_b) {
                return [2 /*return*/, this.service.update(data, where, userId)];
            });
        });
    };
    DishResolver.prototype.createManyDishs = function (_a, userId) {
        var data = _a.data;
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_b) {
                return [2 /*return*/, this.service.createMany(data, userId)];
            });
        });
    };
    DishResolver.prototype.deleteDish = function (where, userId) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.service.delete(where, userId)];
            });
        });
    };
    DishResolver.prototype.successfulTransaction = function (data, ctx) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.service.successfulTransaction(data, ctx.user.id)];
            });
        });
    };
    DishResolver.prototype.failedTransaction = function (data, ctx) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.service.failedTransaction(data, ctx.user.id)];
            });
        });
    };
    __decorate([
        type_graphql_1.Authorized('kitchenSink:read'),
        type_graphql_1.FieldResolver(function () { return kitchen_sink_model_1.KitchenSink; }),
        __param(0, type_graphql_1.Root()), __param(1, type_graphql_1.Ctx()),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [dish_model_1.Dish, Object]),
        __metadata("design:returntype", Promise)
    ], DishResolver.prototype, "kitchenSink", null);
    __decorate([
        type_graphql_1.Authorized('dish:read'),
        type_graphql_1.Query(function () { return [dish_model_1.Dish]; }),
        __param(0, type_graphql_1.Args()),
        __param(1, __1.Fields()),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [generated_1.DishWhereArgs, Array]),
        __metadata("design:returntype", Promise)
    ], DishResolver.prototype, "dishes", null);
    __decorate([
        type_graphql_1.Authorized('dish:read'),
        type_graphql_1.Query(function () { return DishConnection; }),
        __param(0, type_graphql_1.Args()),
        __param(1, __1.RawFields()),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [DishConnectionWhereArgs, Object]),
        __metadata("design:returntype", Promise)
    ], DishResolver.prototype, "dishConnection", null);
    __decorate([
        type_graphql_1.Authorized('dish:read'),
        type_graphql_1.Query(function () { return dish_model_1.Dish; }),
        __param(0, type_graphql_1.Arg('where')),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [generated_1.DishWhereUniqueInput]),
        __metadata("design:returntype", Promise)
    ], DishResolver.prototype, "dish", null);
    __decorate([
        type_graphql_1.Authorized('dish:create'),
        type_graphql_1.Mutation(function () { return dish_model_1.Dish; }),
        __param(0, type_graphql_1.Arg('data')), __param(1, __1.UserId()),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [generated_1.DishCreateInput, String]),
        __metadata("design:returntype", Promise)
    ], DishResolver.prototype, "createDish", null);
    __decorate([
        type_graphql_1.Authorized('dish:update'),
        type_graphql_1.Mutation(function () { return dish_model_1.Dish; }),
        __param(0, type_graphql_1.Args()),
        __param(1, __1.UserId()),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [generated_1.DishUpdateArgs, String]),
        __metadata("design:returntype", Promise)
    ], DishResolver.prototype, "updateDish", null);
    __decorate([
        type_graphql_1.Authorized('dish:create'),
        type_graphql_1.Mutation(function () { return [dish_model_1.Dish]; }),
        __param(0, type_graphql_1.Args()),
        __param(1, __1.UserId()),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [generated_1.DishCreateManyArgs, String]),
        __metadata("design:returntype", Promise)
    ], DishResolver.prototype, "createManyDishs", null);
    __decorate([
        type_graphql_1.Authorized('dish:delete'),
        type_graphql_1.Mutation(function () { return __1.StandardDeleteResponse; }),
        __param(0, type_graphql_1.Arg('where')),
        __param(1, __1.UserId()),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [generated_1.DishWhereUniqueInput, String]),
        __metadata("design:returntype", Promise)
    ], DishResolver.prototype, "deleteDish", null);
    __decorate([
        type_graphql_1.Mutation(function () { return [dish_model_1.Dish]; }),
        __param(0, type_graphql_1.Arg('data')),
        __param(1, type_graphql_1.Ctx()),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [generated_1.DishCreateInput, Object]),
        __metadata("design:returntype", Promise)
    ], DishResolver.prototype, "successfulTransaction", null);
    __decorate([
        type_graphql_1.Mutation(function () { return [dish_model_1.Dish]; }),
        __param(0, type_graphql_1.Arg('data')),
        __param(1, type_graphql_1.Ctx()),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [generated_1.DishCreateInput, Object]),
        __metadata("design:returntype", Promise)
    ], DishResolver.prototype, "failedTransaction", null);
    DishResolver = __decorate([
        type_graphql_1.Resolver(dish_model_1.Dish),
        __param(0, typedi_1.Inject('DishService')),
        __metadata("design:paramtypes", [dish_service_1.DishService])
    ], DishResolver);
    return DishResolver;
}());
exports.DishResolver = DishResolver;
