// @dzlzv/hydra-cli/lib/src/generate/AbstractRenderer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AbstractRenderer = void 0;
const tslib_1 = require("tslib");
const mustache_1 = tslib_1.__importDefault(require("mustache"));
const debug_1 = tslib_1.__importDefault(require("debug"));
const prettier = tslib_1.__importStar(require("prettier"));
const debug = debug_1.default('qnode-cli:abstract-renderer');
class AbstractRenderer {
    constructor(model, context = {}) {
        this.context = {};
        this.context = context;
        this.model = model;
    }
    render(mustacheTeplate) {
        const mustacheContext = this.transform();
        debug(`Rendering with context: ${JSON.stringify(mustacheContext, null, 2)}`);
        const rendered = mustache_1.default.render(mustacheTeplate, mustacheContext);
        return prettier.format(rendered, {
            parser: 'typescript',
            singleQuote: true,
            printWidth: 120,
        });
    }
}
exports.AbstractRenderer = AbstractRenderer;
