// @dzlzv/hydra-cli/lib/src/generate/ConfigProvider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConfigProvider = void 0;
const tslib_1 = require("tslib");
const warthog_1 = require("warthog");
const path = tslib_1.__importStar(require("path"));
const util = tslib_1.__importStar(require("./utils"));
class ConfigProvider {
    constructor() {
        this.config = new warthog_1.Config();
        this.config.loadSync();
        this.cliGeneratedPath = path.join(this.config.get('ROOT_FOLDER'), '/', this.config.get('CLI_GENERATE_PATH'), '/');
    }
    withGeneratedFolderRelPath(name) {
        const destFolder = this.getDestFolder(name);
        const generatedFolderRelPath = path.relative(destFolder, this.config.get('GENERATED_FOLDER'));
        return {
            generatedFolderRelPath,
        };
    }
    getDestFolder(name) {
        return util.supplant(this.cliGeneratedPath, util.names(name));
    }
    getMigrationsFolder() {
        return this.config.get('DB_MIGRATIONS_DIR');
    }
}
exports.ConfigProvider = ConfigProvider;
