// @dzlzv/hydra-cli/lib/src/model/Field.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Field = void 0;
const ScalarTypes_1 = require("./ScalarTypes");
const WarthogModel_1 = require("./WarthogModel");
/**
 * Reperenst GraphQL object type field
 * @constructor(name: string, type: string, nullable: boolean = true, isBuildinType: boolean = true, isList = false)
 */
class Field {
    constructor(name, type, nullable = true, isBuildinType = true, isList = false) {
        this.name = name;
        this.type = type;
        this.nullable = nullable;
        this.isBuildinType = isBuildinType;
        this.isList = isList;
    }
    // get isBuiltInType(): boolean {
    //   return this.modelType == ModelType.SCALAR
    // }
    columnType() {
        var _a;
        if (this.relation)
            return (_a = this.relation) === null || _a === void 0 ? void 0 : _a.type;
        return this.isBuildinType ? ScalarTypes_1.availableTypes[this.type] : this.type;
    }
    isArray() {
        return this.isBuildinType && this.isList;
    }
    isScalar() {
        return this.isBuildinType && !this.isList;
    }
    isRelationType() {
        return !!this.relation;
    }
    isEnum() {
        return this.modelType === WarthogModel_1.ModelType.ENUM;
    }
    isUnion() {
        return this.modelType === WarthogModel_1.ModelType.UNION;
    }
}
exports.Field = Field;
