// @dzlzv/hydra-common/lib/utils/BackOffStrategy.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ExponentialBackOffStrategy = exports.ConstantBackOffStrategy = void 0;
const env_flags_1 = require("./env-flags");
const INITIAL_BACKOFF_TIME = env_flags_1.numberEnv('DEFAULT_BACKOFF_TIME_MS') || 250;
const MAX_BACKOFF_TIME = env_flags_1.numberEnv('MAX_BACKOFF_TIME_MS') || 1000 * 60 * 5; // 5 mins
class ConstantBackOffStrategy {
    constructor(waitTime) {
        this.waitTime = waitTime;
    }
    resetBackoffTime() {
        // NOOP
    }
    registerFailure() {
        // NOOP
    }
    getBackOffMs() {
        return this.waitTime;
    }
}
exports.ConstantBackOffStrategy = ConstantBackOffStrategy;
class ExponentialBackOffStrategy {
    constructor() {
        this.backoffTime = INITIAL_BACKOFF_TIME;
        this.maxBackOffTime = MAX_BACKOFF_TIME;
    }
    resetBackoffTime() {
        this.backoffTime = INITIAL_BACKOFF_TIME;
    }
    registerFailure() {
        this.backoffTime = Math.min(this.maxBackOffTime, Math.ceil(this.backoffTime * 1.2));
    }
    getBackOffMs() {
        return this.backoffTime;
    }
}
exports.ExponentialBackOffStrategy = ExponentialBackOffStrategy;
