// warthog/dist/test/modules/kitchen-sink/kitchen-sink.resolver.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.KitchenSinkResolver = void 0;
var type_graphql_1 = require("type-graphql");
var typedi_1 = require("typedi");
var __1 = require("../../../");
var generated_1 = require("../../generated");
var kitchen_sink_model_1 = require("./kitchen-sink.model");
var kitchen_sink_service_1 = require("./kitchen-sink.service");
var KitchenSinkResolver = /** @class */ (function () {
    function KitchenSinkResolver(service) {
        this.service = service;
    }
    // @Authorized('dish:read')
    KitchenSinkResolver.prototype.dishes = function (kitchenSink, ctx) {
        return ctx.dataLoader.loaders.KitchenSink.dishes.load(kitchenSink);
    };
    KitchenSinkResolver.prototype.kitchenSinks = function (_a, fields) {
        var where = _a.where, orderBy = _a.orderBy, limit = _a.limit, offset = _a.offset;
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_b) {
                return [2 /*return*/, this.service.find(where, orderBy, limit, offset, fields)];
            });
        });
    };
    KitchenSinkResolver.prototype.kitchenSink = function (where) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.service.findOne(where)];
            });
        });
    };
    KitchenSinkResolver.prototype.createKitchenSink = function (data, userId) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.service.create(data, userId)];
            });
        });
    };
    KitchenSinkResolver.prototype.createManyKitchenSinks = function (_a, userId) {
        var data = _a.data;
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_b) {
                return [2 /*return*/, this.service.createMany(data, userId)];
            });
        });
    };
    KitchenSinkResolver.prototype.updateKitchenSink = function (_a, userId) {
        var data = _a.data, where = _a.where;
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_b) {
                return [2 /*return*/, this.service.update(data, where, userId)];
            });
        });
    };
    KitchenSinkResolver.prototype.deleteKitchenSink = function (where, userId) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                return [2 /*return*/, this.service.delete(where, userId)];
            });
        });
    };
    __decorate([
        type_graphql_1.FieldResolver(),
        __param(0, type_graphql_1.Root()), __param(1, type_graphql_1.Ctx()),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [kitchen_sink_model_1.KitchenSink, Object]),
        __metadata("design:returntype", Promise)
    ], KitchenSinkResolver.prototype, "dishes", null);
    __decorate([
        type_graphql_1.Authorized('kitchenSink:read'),
        type_graphql_1.Query(function () { return [kitchen_sink_model_1.KitchenSink]; }),
        __param(0, type_graphql_1.Args()),
        __param(1, __1.Fields()),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [generated_1.KitchenSinkWhereArgs, Array]),
        __metadata("design:returntype", Promise)
    ], KitchenSinkResolver.prototype, "kitchenSinks", null);
    __decorate([
        type_graphql_1.Authorized('kitchenSink:read'),
        type_graphql_1.Query(function () { return kitchen_sink_model_1.KitchenSink; }),
        __param(0, type_graphql_1.Arg('where')),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [generated_1.KitchenSinkWhereUniqueInput]),
        __metadata("design:returntype", Promise)
    ], KitchenSinkResolver.prototype, "kitchenSink", null);
    __decorate([
        type_graphql_1.Authorized('kitchenSink:create'),
        type_graphql_1.Mutation(function () { return kitchen_sink_model_1.KitchenSink; }),
        __param(0, type_graphql_1.Arg('data')),
        __param(1, __1.UserId()),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [generated_1.KitchenSinkCreateInput, String]),
        __metadata("design:returntype", Promise)
    ], KitchenSinkResolver.prototype, "createKitchenSink", null);
    __decorate([
        type_graphql_1.Authorized('kitchenSink:create'),
        type_graphql_1.Mutation(function () { return [kitchen_sink_model_1.KitchenSink]; }),
        __param(0, type_graphql_1.Args()),
        __param(1, __1.UserId()),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [generated_1.KitchenSinkCreateManyArgs, String]),
        __metadata("design:returntype", Promise)
    ], KitchenSinkResolver.prototype, "createManyKitchenSinks", null);
    __decorate([
        type_graphql_1.Authorized('kitchenSink:update'),
        type_graphql_1.Mutation(function () { return kitchen_sink_model_1.KitchenSink; }),
        __param(0, type_graphql_1.Args()),
        __param(1, __1.UserId()),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [generated_1.KitchenSinkUpdateArgs, String]),
        __metadata("design:returntype", Promise)
    ], KitchenSinkResolver.prototype, "updateKitchenSink", null);
    __decorate([
        type_graphql_1.Authorized('kitchenSink:delete'),
        type_graphql_1.Mutation(function () { return __1.StandardDeleteResponse; }),
        __param(0, type_graphql_1.Arg('where')),
        __param(1, __1.UserId()),
        __metadata("design:type", Function),
        __metadata("design:paramtypes", [generated_1.KitchenSinkWhereUniqueInput, String]),
        __metadata("design:returntype", Promise)
    ], KitchenSinkResolver.prototype, "deleteKitchenSink", null);
    KitchenSinkResolver = __decorate([
        type_graphql_1.Resolver(kitchen_sink_model_1.KitchenSink),
        __param(0, typedi_1.Inject('KitchenSinkService')),
        __metadata("design:paramtypes", [kitchen_sink_service_1.KitchenSinkService])
    ], KitchenSinkResolver);
    return KitchenSinkResolver;
}());
exports.KitchenSinkResolver = KitchenSinkResolver;
