// @dzlzv/hydra-processor/lib/prometheus/ProcessorPromClient.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProcessorPromClient = void 0;
const prom_client_1 = require("prom-client");
const typedi_1 = require("typedi");
const hydra_common_1 = require("@dzlzv/hydra-common");
const debug_1 = require("debug");
const db_1 = require("../db");
const debug = debug_1.default('index-builder:processor-prom-client');
class ProcessorPromClient {
    constructor() {
        this.lastScannedBlock = new prom_client_1.Gauge({
            name: 'hydra_processor_last_scanned_block',
            help: 'Last block the processor has scanned for events',
        });
        this.processedEvents = new prom_client_1.Gauge({
            name: 'hydra_processor_processed_events_cnt',
            help: 'total number of processed events',
            labelNames: ['name'],
        });
        typedi_1.default.set('ProcessorPromClient', this);
    }
    init() {
        prom_client_1.collectDefaultMetrics({ prefix: 'hydra_processor_system_' });
        this.stateHandler = typedi_1.default.get('ProcessorStateHandler');
        this.processor = typedi_1.default.get('MappingsProcessor');
        this.initValues()
            .then(() => {
            this.stateHandler.on('STATE_CHANGE', (state) => {
                this.lastScannedBlock.set(state.lastScannedBlock);
            });
            this.processor.on('PROCESSED_EVENT', (event) => {
                this.processedEvents.inc();
                this.processedEvents.inc({ name: event.name });
            });
        })
            .catch((e) => debug(`Error initializing the values: ${hydra_common_1.logError(e)}`));
    }
    async initValues() {
        const totalEvents = await db_1.countProcessedEvents(this.processor.name);
        this.processedEvents.set(totalEvents);
    }
}
exports.ProcessorPromClient = ProcessorPromClient;
