// @dzlzv/hydra-cli/lib/test/helpers/model.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createModel = exports.postObjType = exports.threadObjType = exports.fromStringSchema = void 0;
const tslib_1 = require("tslib");
/* eslint-disable @typescript-eslint/no-unsafe-member-access */
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
/* eslint-disable @typescript-eslint/no-unsafe-call */
const model_1 = require("../../src/model");
const tmp = tslib_1.__importStar(require("tmp"));
const fs = tslib_1.__importStar(require("fs-extra"));
const WarthogModelBuilder_1 = require("../../src/parse/WarthogModelBuilder");
const threadObjType = {
    name: 'Thread',
    isEntity: true,
    isVariant: false,
    fields: [
        new model_1.Field('initial_body_text', 'String'),
        new model_1.Field('title', 'String'),
        new model_1.Field('id', 'ID'),
    ],
};
exports.threadObjType = threadObjType;
const postObjType = {
    name: 'Post',
    isEntity: true,
    isVariant: false,
    fields: [
        new model_1.Field('initial_body_text', 'String'),
        new model_1.Field('title', 'String'),
        new model_1.Field('id', 'ID'),
    ],
};
exports.postObjType = postObjType;
const createModel = () => {
    const warthogModel = new model_1.WarthogModel();
    warthogModel.addEntity(threadObjType);
    warthogModel.addEntity(postObjType);
    return warthogModel;
};
exports.createModel = createModel;
function fromStringSchema(schema) {
    const tmpobj = tmp.fileSync();
    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
    fs.writeSync(tmpobj.fd, schema);
    const modelBuilder = new WarthogModelBuilder_1.WarthogModelBuilder(tmpobj.name);
    return modelBuilder.buildWarthogModel();
}
exports.fromStringSchema = fromStringSchema;
