// warthog/dist/test/generated/classes.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DishUpdateArgs = exports.DishCreateManyArgs = exports.DishWhereArgs = exports.DishUpdateInput = exports.DishCreateInput = exports.DishWhereUniqueInput = exports.DishWhereInput = exports.DishOrderByEnum = exports.KitchenSinkUpdateArgs = exports.KitchenSinkCreateManyArgs = exports.KitchenSinkWhereArgs = exports.KitchenSinkUpdateInput = exports.KitchenSinkCreateInput = exports.KitchenSinkWhereUniqueInput = exports.KitchenSinkWhereInput = exports.KitchenSinkOrderByEnum = exports.ApiOnlyUpdateArgs = exports.ApiOnlyCreateManyArgs = exports.ApiOnlyWhereArgs = exports.ApiOnlyUpdateInput = exports.ApiOnlyCreateInput = exports.ApiOnlyWhereUniqueInput = exports.ApiOnlyWhereInput = exports.ApiOnlyOrderByEnum = void 0;
// This file has been auto-generated by Warthog.  Do not update directly as it
// will be re-written.  If you need to change this file, update models or add
// new TypeGraphQL objects
// prettier-ignore
// @ts-ignore
var graphql_scalars_1 = require("graphql-scalars");
// prettier-ignore
// @ts-ignore
var graphql_1 = require("graphql");
// prettier-ignore
// @ts-ignore
var type_graphql_1 = require("type-graphql");
// prettier-ignore
// @ts-ignore
var type_graphql_2 = require("type-graphql");
// prettier-ignore
// @ts-ignore eslint-disable-next-line @typescript-eslint/no-var-requires
var GraphQLJSONObject = require('graphql-type-json').GraphQLJSONObject;
// prettier-ignore
// @ts-ignore
var __1 = require("../../");
var kitchen_sink_model_1 = require("../modules/kitchen-sink/kitchen-sink.model");
var ApiOnlyOrderByEnum;
(function (ApiOnlyOrderByEnum) {
    ApiOnlyOrderByEnum["createdAt_ASC"] = "createdAt_ASC";
    ApiOnlyOrderByEnum["createdAt_DESC"] = "createdAt_DESC";
    ApiOnlyOrderByEnum["updatedAt_ASC"] = "updatedAt_ASC";
    ApiOnlyOrderByEnum["updatedAt_DESC"] = "updatedAt_DESC";
    ApiOnlyOrderByEnum["deletedAt_ASC"] = "deletedAt_ASC";
    ApiOnlyOrderByEnum["deletedAt_DESC"] = "deletedAt_DESC";
    ApiOnlyOrderByEnum["name_ASC"] = "name_ASC";
    ApiOnlyOrderByEnum["name_DESC"] = "name_DESC";
})(ApiOnlyOrderByEnum = exports.ApiOnlyOrderByEnum || (exports.ApiOnlyOrderByEnum = {}));
type_graphql_2.registerEnumType(ApiOnlyOrderByEnum, {
    name: "ApiOnlyOrderByInput"
});
var ApiOnlyWhereInput = /** @class */ (function () {
    function ApiOnlyWhereInput() {
    }
    __decorate([
        type_graphql_1.Field(function () { return graphql_1.GraphQLID; }, { nullable: true }),
        __metadata("design:type", String)
    ], ApiOnlyWhereInput.prototype, "id_eq", void 0);
    __decorate([
        type_graphql_1.Field(function () { return [graphql_1.GraphQLID]; }, { nullable: true }),
        __metadata("design:type", Array)
    ], ApiOnlyWhereInput.prototype, "id_in", void 0);
    __decorate([
        type_graphql_1.Field(function () { return type_graphql_2.GraphQLISODateTime; }, { nullable: true }),
        __metadata("design:type", Object)
    ], ApiOnlyWhereInput.prototype, "createdAt_eq", void 0);
    __decorate([
        type_graphql_1.Field(function () { return type_graphql_2.GraphQLISODateTime; }, { nullable: true }),
        __metadata("design:type", Object)
    ], ApiOnlyWhereInput.prototype, "createdAt_lt", void 0);
    __decorate([
        type_graphql_1.Field(function () { return type_graphql_2.GraphQLISODateTime; }, { nullable: true }),
        __metadata("design:type", Object)
    ], ApiOnlyWhereInput.prototype, "createdAt_lte", void 0);
    __decorate([
        type_graphql_1.Field(function () { return type_graphql_2.GraphQLISODateTime; }, { nullable: true }),
        __metadata("design:type", Object)
    ], ApiOnlyWhereInput.prototype, "createdAt_gt", void 0);
    __decorate([
        type_graphql_1.Field(function () { return type_graphql_2.GraphQLISODateTime; }, { nullable: true }),
        __metadata("design:type", Object)
    ], ApiOnlyWhereInput.prototype, "createdAt_gte", void 0);
    __decorate([
        type_graphql_1.Field(function () { return graphql_1.GraphQLID; }, { nullable: true }),
        __metadata("design:type", String)
    ], ApiOnlyWhereInput.prototype, "createdById_eq", void 0);
    __decorate([
        type_graphql_1.Field(function () { return [graphql_1.GraphQLID]; }, { nullable: true }),
        __metadata("design:type", Array)
    ], ApiOnlyWhereInput.prototype, "createdById_in", void 0);
    __decorate([
        type_graphql_1.Field(function () { return type_graphql_2.GraphQLISODateTime; }, { nullable: true }),
        __metadata("design:type", Object)
    ], ApiOnlyWhereInput.prototype, "updatedAt_eq", void 0);
    __decorate([
        type_graphql_1.Field(function () { return type_graphql_2.GraphQLISODateTime; }, { nullable: true }),
        __metadata("design:type", Object)
    ], ApiOnlyWhereInput.prototype, "updatedAt_lt", void 0);
    __decorate([
        type_graphql_1.Field(function () { return type_graphql_2.GraphQLISODateTime; }, { nullable: true }),
        __metadata("design:type", Object)
    ], ApiOnlyWhereInput.prototype, "updatedAt_lte", void 0);
    __decorate([
        type_graphql_1.Field(function () { return type_graphql_2.GraphQLISODateTime; }, { nullable: true }),
        __metadata("design:type", Object)
    ], ApiOnlyWhereInput.prototype, "updatedAt_gt", void 0);
    __decorate([
        type_graphql_1.Field(function () { return type_graphql_2.GraphQLISODateTime; }, { nullable: true }),
        __metadata("design:type", Object)
    ], ApiOnlyWhereInput.prototype, "updatedAt_gte", void 0);
    __decorate([
        type_graphql_1.Field(function () { return graphql_1.GraphQLID; }, { nullable: true }),
        __metadata("design:type", String)
    ], ApiOnlyWhereInput.prototype, "updatedById_eq", void 0);
    __decorate([
        type_graphql_1.Field(function () { return [graphql_1.GraphQLID]; }, { nullable: true }),
        __metadata("design:type", Array)
    ], ApiOnlyWhereInput.prototype, "updatedById_in", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", Boolean)
    ], ApiOnlyWhereInput.prototype, "deletedAt_all", void 0);
    __decorate([
        type_graphql_1.Field(function () { return type_graphql_2.GraphQLISODateTime; }, { nullable: true }),
        __metadata("design:type", Object)
    ], ApiOnlyWhereInput.prototype, "deletedAt_eq", void 0);
    __decorate([
        type_graphql_1.Field(function () { return type_graphql_2.GraphQLISODateTime; }, { nullable: true }),
        __metadata("design:type", Object)
    ], ApiOnlyWhereInput.prototype, "deletedAt_lt", void 0);
    __decorate([
        type_graphql_1.Field(function () { return type_graphql_2.GraphQLISODateTime; }, { nullable: true }),
        __metadata("design:type", Object)
    ], ApiOnlyWhereInput.prototype, "deletedAt_lte", void 0);
    __decorate([
        type_graphql_1.Field(function () { return type_graphql_2.GraphQLISODateTime; }, { nullable: true }),
        __metadata("design:type", Object)
    ], ApiOnlyWhereInput.prototype, "deletedAt_gt", void 0);
    __decorate([
        type_graphql_1.Field(function () { return type_graphql_2.GraphQLISODateTime; }, { nullable: true }),
        __metadata("design:type", Object)
    ], ApiOnlyWhereInput.prototype, "deletedAt_gte", void 0);
    __decorate([
        type_graphql_1.Field(function () { return graphql_1.GraphQLID; }, { nullable: true }),
        __metadata("design:type", String)
    ], ApiOnlyWhereInput.prototype, "deletedById_eq", void 0);
    __decorate([
        type_graphql_1.Field(function () { return [graphql_1.GraphQLID]; }, { nullable: true }),
        __metadata("design:type", Array)
    ], ApiOnlyWhereInput.prototype, "deletedById_in", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], ApiOnlyWhereInput.prototype, "name_eq", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], ApiOnlyWhereInput.prototype, "name_contains", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], ApiOnlyWhereInput.prototype, "name_startsWith", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], ApiOnlyWhereInput.prototype, "name_endsWith", void 0);
    __decorate([
        type_graphql_1.Field(function () { return [String]; }, { nullable: true }),
        __metadata("design:type", Array)
    ], ApiOnlyWhereInput.prototype, "name_in", void 0);
    ApiOnlyWhereInput = __decorate([
        type_graphql_1.InputType()
    ], ApiOnlyWhereInput);
    return ApiOnlyWhereInput;
}());
exports.ApiOnlyWhereInput = ApiOnlyWhereInput;
var ApiOnlyWhereUniqueInput = /** @class */ (function () {
    function ApiOnlyWhereUniqueInput() {
    }
    __decorate([
        type_graphql_1.Field(function () { return graphql_1.GraphQLID; }),
        __metadata("design:type", String)
    ], ApiOnlyWhereUniqueInput.prototype, "id", void 0);
    ApiOnlyWhereUniqueInput = __decorate([
        type_graphql_1.InputType()
    ], ApiOnlyWhereUniqueInput);
    return ApiOnlyWhereUniqueInput;
}());
exports.ApiOnlyWhereUniqueInput = ApiOnlyWhereUniqueInput;
var ApiOnlyCreateInput = /** @class */ (function () {
    function ApiOnlyCreateInput() {
    }
    __decorate([
        type_graphql_1.Field(),
        __metadata("design:type", String)
    ], ApiOnlyCreateInput.prototype, "name", void 0);
    ApiOnlyCreateInput = __decorate([
        type_graphql_1.InputType()
    ], ApiOnlyCreateInput);
    return ApiOnlyCreateInput;
}());
exports.ApiOnlyCreateInput = ApiOnlyCreateInput;
var ApiOnlyUpdateInput = /** @class */ (function () {
    function ApiOnlyUpdateInput() {
    }
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], ApiOnlyUpdateInput.prototype, "name", void 0);
    ApiOnlyUpdateInput = __decorate([
        type_graphql_1.InputType()
    ], ApiOnlyUpdateInput);
    return ApiOnlyUpdateInput;
}());
exports.ApiOnlyUpdateInput = ApiOnlyUpdateInput;
var ApiOnlyWhereArgs = /** @class */ (function (_super) {
    __extends(ApiOnlyWhereArgs, _super);
    function ApiOnlyWhereArgs() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    __decorate([
        type_graphql_1.Field(function () { return ApiOnlyWhereInput; }, { nullable: true }),
        __metadata("design:type", ApiOnlyWhereInput)
    ], ApiOnlyWhereArgs.prototype, "where", void 0);
    __decorate([
        type_graphql_1.Field(function () { return ApiOnlyOrderByEnum; }, { nullable: true }),
        __metadata("design:type", String)
    ], ApiOnlyWhereArgs.prototype, "orderBy", void 0);
    ApiOnlyWhereArgs = __decorate([
        type_graphql_1.ArgsType()
    ], ApiOnlyWhereArgs);
    return ApiOnlyWhereArgs;
}(__1.PaginationArgs));
exports.ApiOnlyWhereArgs = ApiOnlyWhereArgs;
var ApiOnlyCreateManyArgs = /** @class */ (function () {
    function ApiOnlyCreateManyArgs() {
    }
    __decorate([
        type_graphql_1.Field(function () { return [ApiOnlyCreateInput]; }),
        __metadata("design:type", Array)
    ], ApiOnlyCreateManyArgs.prototype, "data", void 0);
    ApiOnlyCreateManyArgs = __decorate([
        type_graphql_1.ArgsType()
    ], ApiOnlyCreateManyArgs);
    return ApiOnlyCreateManyArgs;
}());
exports.ApiOnlyCreateManyArgs = ApiOnlyCreateManyArgs;
var ApiOnlyUpdateArgs = /** @class */ (function () {
    function ApiOnlyUpdateArgs() {
    }
    __decorate([
        type_graphql_1.Field(),
        __metadata("design:type", ApiOnlyUpdateInput)
    ], ApiOnlyUpdateArgs.prototype, "data", void 0);
    __decorate([
        type_graphql_1.Field(),
        __metadata("design:type", ApiOnlyWhereUniqueInput)
    ], ApiOnlyUpdateArgs.prototype, "where", void 0);
    ApiOnlyUpdateArgs = __decorate([
        type_graphql_1.ArgsType()
    ], ApiOnlyUpdateArgs);
    return ApiOnlyUpdateArgs;
}());
exports.ApiOnlyUpdateArgs = ApiOnlyUpdateArgs;
var KitchenSinkOrderByEnum;
(function (KitchenSinkOrderByEnum) {
    KitchenSinkOrderByEnum["createdAt_ASC"] = "createdAt_ASC";
    KitchenSinkOrderByEnum["createdAt_DESC"] = "createdAt_DESC";
    KitchenSinkOrderByEnum["updatedAt_ASC"] = "updatedAt_ASC";
    KitchenSinkOrderByEnum["updatedAt_DESC"] = "updatedAt_DESC";
    KitchenSinkOrderByEnum["deletedAt_ASC"] = "deletedAt_ASC";
    KitchenSinkOrderByEnum["deletedAt_DESC"] = "deletedAt_DESC";
    KitchenSinkOrderByEnum["stringField_ASC"] = "stringField_ASC";
    KitchenSinkOrderByEnum["stringField_DESC"] = "stringField_DESC";
    KitchenSinkOrderByEnum["nullableStringField_ASC"] = "nullableStringField_ASC";
    KitchenSinkOrderByEnum["nullableStringField_DESC"] = "nullableStringField_DESC";
    KitchenSinkOrderByEnum["dateField_ASC"] = "dateField_ASC";
    KitchenSinkOrderByEnum["dateField_DESC"] = "dateField_DESC";
    KitchenSinkOrderByEnum["dateOnlyField_ASC"] = "dateOnlyField_ASC";
    KitchenSinkOrderByEnum["dateOnlyField_DESC"] = "dateOnlyField_DESC";
    KitchenSinkOrderByEnum["dateTimeField_ASC"] = "dateTimeField_ASC";
    KitchenSinkOrderByEnum["dateTimeField_DESC"] = "dateTimeField_DESC";
    KitchenSinkOrderByEnum["emailField_ASC"] = "emailField_ASC";
    KitchenSinkOrderByEnum["emailField_DESC"] = "emailField_DESC";
    KitchenSinkOrderByEnum["integerField_ASC"] = "integerField_ASC";
    KitchenSinkOrderByEnum["integerField_DESC"] = "integerField_DESC";
    KitchenSinkOrderByEnum["booleanField_ASC"] = "booleanField_ASC";
    KitchenSinkOrderByEnum["booleanField_DESC"] = "booleanField_DESC";
    KitchenSinkOrderByEnum["floatField_ASC"] = "floatField_ASC";
    KitchenSinkOrderByEnum["floatField_DESC"] = "floatField_DESC";
    KitchenSinkOrderByEnum["idField_ASC"] = "idField_ASC";
    KitchenSinkOrderByEnum["idField_DESC"] = "idField_DESC";
    KitchenSinkOrderByEnum["stringEnumField_ASC"] = "stringEnumField_ASC";
    KitchenSinkOrderByEnum["stringEnumField_DESC"] = "stringEnumField_DESC";
    KitchenSinkOrderByEnum["numericField_ASC"] = "numericField_ASC";
    KitchenSinkOrderByEnum["numericField_DESC"] = "numericField_DESC";
    KitchenSinkOrderByEnum["numericFieldCustomPrecisionScale_ASC"] = "numericFieldCustomPrecisionScale_ASC";
    KitchenSinkOrderByEnum["numericFieldCustomPrecisionScale_DESC"] = "numericFieldCustomPrecisionScale_DESC";
    KitchenSinkOrderByEnum["noFilterField_ASC"] = "noFilterField_ASC";
    KitchenSinkOrderByEnum["noFilterField_DESC"] = "noFilterField_DESC";
    KitchenSinkOrderByEnum["characterField_ASC"] = "characterField_ASC";
    KitchenSinkOrderByEnum["characterField_DESC"] = "characterField_DESC";
    KitchenSinkOrderByEnum["readonlyField_ASC"] = "readonlyField_ASC";
    KitchenSinkOrderByEnum["readonlyField_DESC"] = "readonlyField_DESC";
    KitchenSinkOrderByEnum["apiOnlyField_ASC"] = "apiOnlyField_ASC";
    KitchenSinkOrderByEnum["apiOnlyField_DESC"] = "apiOnlyField_DESC";
})(KitchenSinkOrderByEnum = exports.KitchenSinkOrderByEnum || (exports.KitchenSinkOrderByEnum = {}));
type_graphql_2.registerEnumType(KitchenSinkOrderByEnum, {
    name: "KitchenSinkOrderByInput"
});
var KitchenSinkWhereInput = /** @class */ (function () {
    function KitchenSinkWhereInput() {
    }
    __decorate([
        type_graphql_1.Field(function () { return graphql_1.GraphQLID; }, { nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkWhereInput.prototype, "id_eq", void 0);
    __decorate([
        type_graphql_1.Field(function () { return [graphql_1.GraphQLID]; }, { nullable: true }),
        __metadata("design:type", Array)
    ], KitchenSinkWhereInput.prototype, "id_in", void 0);
    __decorate([
        type_graphql_1.Field(function () { return type_graphql_2.GraphQLISODateTime; }, { nullable: true }),
        __metadata("design:type", Object)
    ], KitchenSinkWhereInput.prototype, "createdAt_eq", void 0);
    __decorate([
        type_graphql_1.Field(function () { return type_graphql_2.GraphQLISODateTime; }, { nullable: true }),
        __metadata("design:type", Object)
    ], KitchenSinkWhereInput.prototype, "createdAt_lt", void 0);
    __decorate([
        type_graphql_1.Field(function () { return type_graphql_2.GraphQLISODateTime; }, { nullable: true }),
        __metadata("design:type", Object)
    ], KitchenSinkWhereInput.prototype, "createdAt_lte", void 0);
    __decorate([
        type_graphql_1.Field(function () { return type_graphql_2.GraphQLISODateTime; }, { nullable: true }),
        __metadata("design:type", Object)
    ], KitchenSinkWhereInput.prototype, "createdAt_gt", void 0);
    __decorate([
        type_graphql_1.Field(function () { return type_graphql_2.GraphQLISODateTime; }, { nullable: true }),
        __metadata("design:type", Object)
    ], KitchenSinkWhereInput.prototype, "createdAt_gte", void 0);
    __decorate([
        type_graphql_1.Field(function () { return graphql_1.GraphQLID; }, { nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkWhereInput.prototype, "createdById_eq", void 0);
    __decorate([
        type_graphql_1.Field(function () { return [graphql_1.GraphQLID]; }, { nullable: true }),
        __metadata("design:type", Array)
    ], KitchenSinkWhereInput.prototype, "createdById_in", void 0);
    __decorate([
        type_graphql_1.Field(function () { return type_graphql_2.GraphQLISODateTime; }, { nullable: true }),
        __metadata("design:type", Object)
    ], KitchenSinkWhereInput.prototype, "updatedAt_eq", void 0);
    __decorate([
        type_graphql_1.Field(function () { return type_graphql_2.GraphQLISODateTime; }, { nullable: true }),
        __metadata("design:type", Object)
    ], KitchenSinkWhereInput.prototype, "updatedAt_lt", void 0);
    __decorate([
        type_graphql_1.Field(function () { return type_graphql_2.GraphQLISODateTime; }, { nullable: true }),
        __metadata("design:type", Object)
    ], KitchenSinkWhereInput.prototype, "updatedAt_lte", void 0);
    __decorate([
        type_graphql_1.Field(function () { return type_graphql_2.GraphQLISODateTime; }, { nullable: true }),
        __metadata("design:type", Object)
    ], KitchenSinkWhereInput.prototype, "updatedAt_gt", void 0);
    __decorate([
        type_graphql_1.Field(function () { return type_graphql_2.GraphQLISODateTime; }, { nullable: true }),
        __metadata("design:type", Object)
    ], KitchenSinkWhereInput.prototype, "updatedAt_gte", void 0);
    __decorate([
        type_graphql_1.Field(function () { return graphql_1.GraphQLID; }, { nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkWhereInput.prototype, "updatedById_eq", void 0);
    __decorate([
        type_graphql_1.Field(function () { return [graphql_1.GraphQLID]; }, { nullable: true }),
        __metadata("design:type", Array)
    ], KitchenSinkWhereInput.prototype, "updatedById_in", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", Boolean)
    ], KitchenSinkWhereInput.prototype, "deletedAt_all", void 0);
    __decorate([
        type_graphql_1.Field(function () { return type_graphql_2.GraphQLISODateTime; }, { nullable: true }),
        __metadata("design:type", Object)
    ], KitchenSinkWhereInput.prototype, "deletedAt_eq", void 0);
    __decorate([
        type_graphql_1.Field(function () { return type_graphql_2.GraphQLISODateTime; }, { nullable: true }),
        __metadata("design:type", Object)
    ], KitchenSinkWhereInput.prototype, "deletedAt_lt", void 0);
    __decorate([
        type_graphql_1.Field(function () { return type_graphql_2.GraphQLISODateTime; }, { nullable: true }),
        __metadata("design:type", Object)
    ], KitchenSinkWhereInput.prototype, "deletedAt_lte", void 0);
    __decorate([
        type_graphql_1.Field(function () { return type_graphql_2.GraphQLISODateTime; }, { nullable: true }),
        __metadata("design:type", Object)
    ], KitchenSinkWhereInput.prototype, "deletedAt_gt", void 0);
    __decorate([
        type_graphql_1.Field(function () { return type_graphql_2.GraphQLISODateTime; }, { nullable: true }),
        __metadata("design:type", Object)
    ], KitchenSinkWhereInput.prototype, "deletedAt_gte", void 0);
    __decorate([
        type_graphql_1.Field(function () { return graphql_1.GraphQLID; }, { nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkWhereInput.prototype, "deletedById_eq", void 0);
    __decorate([
        type_graphql_1.Field(function () { return [graphql_1.GraphQLID]; }, { nullable: true }),
        __metadata("design:type", Array)
    ], KitchenSinkWhereInput.prototype, "deletedById_in", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkWhereInput.prototype, "stringField_eq", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkWhereInput.prototype, "stringField_contains", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkWhereInput.prototype, "stringField_startsWith", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkWhereInput.prototype, "stringField_endsWith", void 0);
    __decorate([
        type_graphql_1.Field(function () { return [String]; }, { nullable: true }),
        __metadata("design:type", Array)
    ], KitchenSinkWhereInput.prototype, "stringField_in", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkWhereInput.prototype, "nullableStringField_eq", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkWhereInput.prototype, "nullableStringField_contains", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkWhereInput.prototype, "nullableStringField_startsWith", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkWhereInput.prototype, "nullableStringField_endsWith", void 0);
    __decorate([
        type_graphql_1.Field(function () { return [String]; }, { nullable: true }),
        __metadata("design:type", Array)
    ], KitchenSinkWhereInput.prototype, "nullableStringField_in", void 0);
    __decorate([
        type_graphql_1.Field(function () { return type_graphql_2.GraphQLISODateTime; }, { nullable: true }),
        __metadata("design:type", Object)
    ], KitchenSinkWhereInput.prototype, "dateField_eq", void 0);
    __decorate([
        type_graphql_1.Field(function () { return type_graphql_2.GraphQLISODateTime; }, { nullable: true }),
        __metadata("design:type", Object)
    ], KitchenSinkWhereInput.prototype, "dateField_lt", void 0);
    __decorate([
        type_graphql_1.Field(function () { return type_graphql_2.GraphQLISODateTime; }, { nullable: true }),
        __metadata("design:type", Object)
    ], KitchenSinkWhereInput.prototype, "dateField_lte", void 0);
    __decorate([
        type_graphql_1.Field(function () { return type_graphql_2.GraphQLISODateTime; }, { nullable: true }),
        __metadata("design:type", Object)
    ], KitchenSinkWhereInput.prototype, "dateField_gt", void 0);
    __decorate([
        type_graphql_1.Field(function () { return type_graphql_2.GraphQLISODateTime; }, { nullable: true }),
        __metadata("design:type", Object)
    ], KitchenSinkWhereInput.prototype, "dateField_gte", void 0);
    __decorate([
        type_graphql_1.Field(function () { return graphql_scalars_1.DateResolver; }, { nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkWhereInput.prototype, "dateOnlyField_eq", void 0);
    __decorate([
        type_graphql_1.Field(function () { return graphql_scalars_1.DateResolver; }, { nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkWhereInput.prototype, "dateOnlyField_lt", void 0);
    __decorate([
        type_graphql_1.Field(function () { return graphql_scalars_1.DateResolver; }, { nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkWhereInput.prototype, "dateOnlyField_lte", void 0);
    __decorate([
        type_graphql_1.Field(function () { return graphql_scalars_1.DateResolver; }, { nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkWhereInput.prototype, "dateOnlyField_gt", void 0);
    __decorate([
        type_graphql_1.Field(function () { return graphql_scalars_1.DateResolver; }, { nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkWhereInput.prototype, "dateOnlyField_gte", void 0);
    __decorate([
        type_graphql_1.Field(function () { return type_graphql_2.GraphQLISODateTime; }, { nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkWhereInput.prototype, "dateTimeField_eq", void 0);
    __decorate([
        type_graphql_1.Field(function () { return type_graphql_2.GraphQLISODateTime; }, { nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkWhereInput.prototype, "dateTimeField_lt", void 0);
    __decorate([
        type_graphql_1.Field(function () { return type_graphql_2.GraphQLISODateTime; }, { nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkWhereInput.prototype, "dateTimeField_lte", void 0);
    __decorate([
        type_graphql_1.Field(function () { return type_graphql_2.GraphQLISODateTime; }, { nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkWhereInput.prototype, "dateTimeField_gt", void 0);
    __decorate([
        type_graphql_1.Field(function () { return type_graphql_2.GraphQLISODateTime; }, { nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkWhereInput.prototype, "dateTimeField_gte", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkWhereInput.prototype, "emailField_eq", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkWhereInput.prototype, "emailField_contains", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkWhereInput.prototype, "emailField_startsWith", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkWhereInput.prototype, "emailField_endsWith", void 0);
    __decorate([
        type_graphql_1.Field(function () { return [String]; }, { nullable: true }),
        __metadata("design:type", Array)
    ], KitchenSinkWhereInput.prototype, "emailField_in", void 0);
    __decorate([
        type_graphql_1.Field(function () { return type_graphql_1.Int; }, { nullable: true }),
        __metadata("design:type", Number)
    ], KitchenSinkWhereInput.prototype, "integerField_eq", void 0);
    __decorate([
        type_graphql_1.Field(function () { return type_graphql_1.Int; }, { nullable: true }),
        __metadata("design:type", Number)
    ], KitchenSinkWhereInput.prototype, "integerField_gt", void 0);
    __decorate([
        type_graphql_1.Field(function () { return type_graphql_1.Int; }, { nullable: true }),
        __metadata("design:type", Number)
    ], KitchenSinkWhereInput.prototype, "integerField_gte", void 0);
    __decorate([
        type_graphql_1.Field(function () { return type_graphql_1.Int; }, { nullable: true }),
        __metadata("design:type", Number)
    ], KitchenSinkWhereInput.prototype, "integerField_lt", void 0);
    __decorate([
        type_graphql_1.Field(function () { return type_graphql_1.Int; }, { nullable: true }),
        __metadata("design:type", Number)
    ], KitchenSinkWhereInput.prototype, "integerField_lte", void 0);
    __decorate([
        type_graphql_1.Field(function () { return [type_graphql_1.Int]; }, { nullable: true }),
        __metadata("design:type", Array)
    ], KitchenSinkWhereInput.prototype, "integerField_in", void 0);
    __decorate([
        type_graphql_1.Field(function () { return Boolean; }, { nullable: true }),
        __metadata("design:type", Boolean)
    ], KitchenSinkWhereInput.prototype, "booleanField_eq", void 0);
    __decorate([
        type_graphql_1.Field(function () { return [Boolean]; }, { nullable: true }),
        __metadata("design:type", Array)
    ], KitchenSinkWhereInput.prototype, "booleanField_in", void 0);
    __decorate([
        type_graphql_1.Field(function () { return type_graphql_1.Float; }, { nullable: true }),
        __metadata("design:type", Number)
    ], KitchenSinkWhereInput.prototype, "floatField_eq", void 0);
    __decorate([
        type_graphql_1.Field(function () { return type_graphql_1.Float; }, { nullable: true }),
        __metadata("design:type", Number)
    ], KitchenSinkWhereInput.prototype, "floatField_gt", void 0);
    __decorate([
        type_graphql_1.Field(function () { return type_graphql_1.Float; }, { nullable: true }),
        __metadata("design:type", Number)
    ], KitchenSinkWhereInput.prototype, "floatField_gte", void 0);
    __decorate([
        type_graphql_1.Field(function () { return type_graphql_1.Float; }, { nullable: true }),
        __metadata("design:type", Number)
    ], KitchenSinkWhereInput.prototype, "floatField_lt", void 0);
    __decorate([
        type_graphql_1.Field(function () { return type_graphql_1.Float; }, { nullable: true }),
        __metadata("design:type", Number)
    ], KitchenSinkWhereInput.prototype, "floatField_lte", void 0);
    __decorate([
        type_graphql_1.Field(function () { return [type_graphql_1.Float]; }, { nullable: true }),
        __metadata("design:type", Array)
    ], KitchenSinkWhereInput.prototype, "floatField_in", void 0);
    __decorate([
        type_graphql_1.Field(function () { return GraphQLJSONObject; }, { nullable: true }),
        __metadata("design:type", Object)
    ], KitchenSinkWhereInput.prototype, "jsonField_json", void 0);
    __decorate([
        type_graphql_1.Field(function () { return graphql_1.GraphQLID; }, { nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkWhereInput.prototype, "idField_eq", void 0);
    __decorate([
        type_graphql_1.Field(function () { return [graphql_1.GraphQLID]; }, { nullable: true }),
        __metadata("design:type", Array)
    ], KitchenSinkWhereInput.prototype, "idField_in", void 0);
    __decorate([
        type_graphql_1.Field(function () { return kitchen_sink_model_1.StringEnum; }, { nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkWhereInput.prototype, "stringEnumField_eq", void 0);
    __decorate([
        type_graphql_1.Field(function () { return [kitchen_sink_model_1.StringEnum]; }, { nullable: true }),
        __metadata("design:type", Array)
    ], KitchenSinkWhereInput.prototype, "stringEnumField_in", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkWhereInput.prototype, "numericField_eq", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkWhereInput.prototype, "numericField_contains", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkWhereInput.prototype, "numericField_startsWith", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkWhereInput.prototype, "numericField_endsWith", void 0);
    __decorate([
        type_graphql_1.Field(function () { return [String]; }, { nullable: true }),
        __metadata("design:type", Array)
    ], KitchenSinkWhereInput.prototype, "numericField_in", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkWhereInput.prototype, "numericFieldCustomPrecisionScale_eq", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkWhereInput.prototype, "numericFieldCustomPrecisionScale_contains", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkWhereInput.prototype, "numericFieldCustomPrecisionScale_startsWith", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkWhereInput.prototype, "numericFieldCustomPrecisionScale_endsWith", void 0);
    __decorate([
        type_graphql_1.Field(function () { return [String]; }, { nullable: true }),
        __metadata("design:type", Array)
    ], KitchenSinkWhereInput.prototype, "numericFieldCustomPrecisionScale_in", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkWhereInput.prototype, "noSortField_eq", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkWhereInput.prototype, "noSortField_contains", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkWhereInput.prototype, "noSortField_startsWith", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkWhereInput.prototype, "noSortField_endsWith", void 0);
    __decorate([
        type_graphql_1.Field(function () { return [String]; }, { nullable: true }),
        __metadata("design:type", Array)
    ], KitchenSinkWhereInput.prototype, "noSortField_in", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkWhereInput.prototype, "stringFieldFilterEqContains_eq", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkWhereInput.prototype, "stringFieldFilterEqContains_contains", void 0);
    __decorate([
        type_graphql_1.Field(function () { return type_graphql_1.Int; }, { nullable: true }),
        __metadata("design:type", Number)
    ], KitchenSinkWhereInput.prototype, "intFieldFilterLteGte_gte", void 0);
    __decorate([
        type_graphql_1.Field(function () { return type_graphql_1.Int; }, { nullable: true }),
        __metadata("design:type", Number)
    ], KitchenSinkWhereInput.prototype, "intFieldFilterLteGte_lte", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkWhereInput.prototype, "characterField_eq", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkWhereInput.prototype, "characterField_contains", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkWhereInput.prototype, "characterField_startsWith", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkWhereInput.prototype, "characterField_endsWith", void 0);
    __decorate([
        type_graphql_1.Field(function () { return [String]; }, { nullable: true }),
        __metadata("design:type", Array)
    ], KitchenSinkWhereInput.prototype, "characterField_in", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkWhereInput.prototype, "readonlyField_eq", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkWhereInput.prototype, "readonlyField_contains", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkWhereInput.prototype, "readonlyField_startsWith", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkWhereInput.prototype, "readonlyField_endsWith", void 0);
    __decorate([
        type_graphql_1.Field(function () { return [String]; }, { nullable: true }),
        __metadata("design:type", Array)
    ], KitchenSinkWhereInput.prototype, "readonlyField_in", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkWhereInput.prototype, "apiOnlyField_eq", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkWhereInput.prototype, "apiOnlyField_contains", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkWhereInput.prototype, "apiOnlyField_startsWith", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkWhereInput.prototype, "apiOnlyField_endsWith", void 0);
    __decorate([
        type_graphql_1.Field(function () { return [String]; }, { nullable: true }),
        __metadata("design:type", Array)
    ], KitchenSinkWhereInput.prototype, "apiOnlyField_in", void 0);
    __decorate([
        type_graphql_1.Field(function () { return [String]; }, { nullable: true }),
        __metadata("design:type", Array)
    ], KitchenSinkWhereInput.prototype, "arrayOfStrings_containsAll", void 0);
    __decorate([
        type_graphql_1.Field(function () { return [String]; }, { nullable: true }),
        __metadata("design:type", Array)
    ], KitchenSinkWhereInput.prototype, "arrayOfStrings_containsNone", void 0);
    __decorate([
        type_graphql_1.Field(function () { return [String]; }, { nullable: true }),
        __metadata("design:type", Array)
    ], KitchenSinkWhereInput.prototype, "arrayOfStrings_containsAny", void 0);
    __decorate([
        type_graphql_1.Field(function () { return [type_graphql_1.Int]; }, { nullable: true }),
        __metadata("design:type", Array)
    ], KitchenSinkWhereInput.prototype, "arrayOfInts_containsAll", void 0);
    __decorate([
        type_graphql_1.Field(function () { return [type_graphql_1.Int]; }, { nullable: true }),
        __metadata("design:type", Array)
    ], KitchenSinkWhereInput.prototype, "arrayOfInts_containsNone", void 0);
    __decorate([
        type_graphql_1.Field(function () { return [type_graphql_1.Int]; }, { nullable: true }),
        __metadata("design:type", Array)
    ], KitchenSinkWhereInput.prototype, "arrayOfInts_containsAny", void 0);
    KitchenSinkWhereInput = __decorate([
        type_graphql_1.InputType()
    ], KitchenSinkWhereInput);
    return KitchenSinkWhereInput;
}());
exports.KitchenSinkWhereInput = KitchenSinkWhereInput;
var KitchenSinkWhereUniqueInput = /** @class */ (function () {
    function KitchenSinkWhereUniqueInput() {
    }
    __decorate([
        type_graphql_1.Field(function () { return graphql_1.GraphQLID; }, { nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkWhereUniqueInput.prototype, "id", void 0);
    __decorate([
        type_graphql_1.Field(function () { return String; }, { nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkWhereUniqueInput.prototype, "emailField", void 0);
    KitchenSinkWhereUniqueInput = __decorate([
        type_graphql_1.InputType()
    ], KitchenSinkWhereUniqueInput);
    return KitchenSinkWhereUniqueInput;
}());
exports.KitchenSinkWhereUniqueInput = KitchenSinkWhereUniqueInput;
var KitchenSinkCreateInput = /** @class */ (function () {
    function KitchenSinkCreateInput() {
    }
    __decorate([
        type_graphql_1.Field(),
        __metadata("design:type", String)
    ], KitchenSinkCreateInput.prototype, "stringField", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkCreateInput.prototype, "nullableStringField", void 0);
    __decorate([
        type_graphql_1.Field(function () { return type_graphql_2.GraphQLISODateTime; }, { nullable: true }),
        __metadata("design:type", Object)
    ], KitchenSinkCreateInput.prototype, "dateField", void 0);
    __decorate([
        type_graphql_1.Field(function () { return graphql_scalars_1.DateResolver; }, { nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkCreateInput.prototype, "dateOnlyField", void 0);
    __decorate([
        type_graphql_1.Field(function () { return type_graphql_2.GraphQLISODateTime; }, { nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkCreateInput.prototype, "dateTimeField", void 0);
    __decorate([
        type_graphql_1.Field(),
        __metadata("design:type", String)
    ], KitchenSinkCreateInput.prototype, "emailField", void 0);
    __decorate([
        type_graphql_1.Field(),
        __metadata("design:type", Number)
    ], KitchenSinkCreateInput.prototype, "integerField", void 0);
    __decorate([
        type_graphql_1.Field(),
        __metadata("design:type", Boolean)
    ], KitchenSinkCreateInput.prototype, "booleanField", void 0);
    __decorate([
        type_graphql_1.Field(),
        __metadata("design:type", Number)
    ], KitchenSinkCreateInput.prototype, "floatField", void 0);
    __decorate([
        type_graphql_1.Field(function () { return GraphQLJSONObject; }, { nullable: true }),
        __metadata("design:type", Object)
    ], KitchenSinkCreateInput.prototype, "jsonField", void 0);
    __decorate([
        type_graphql_1.Field(function () { return graphql_1.GraphQLID; }, { nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkCreateInput.prototype, "idField", void 0);
    __decorate([
        type_graphql_1.Field(function () { return kitchen_sink_model_1.StringEnum; }, { nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkCreateInput.prototype, "stringEnumField", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkCreateInput.prototype, "numericField", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkCreateInput.prototype, "numericFieldCustomPrecisionScale", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkCreateInput.prototype, "noFilterField", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkCreateInput.prototype, "noSortField", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkCreateInput.prototype, "noFilterOrSortField", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkCreateInput.prototype, "stringFieldFilterEqContains", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", Number)
    ], KitchenSinkCreateInput.prototype, "intFieldFilterLteGte", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkCreateInput.prototype, "characterField", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkCreateInput.prototype, "customTextFieldNoSortOrFilter", void 0);
    __decorate([
        type_graphql_1.Field(function () { return [String]; }, { nullable: true }),
        __metadata("design:type", Array)
    ], KitchenSinkCreateInput.prototype, "customFieldArrayColumn", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkCreateInput.prototype, "writeonlyField", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkCreateInput.prototype, "apiOnlyField", void 0);
    __decorate([
        type_graphql_1.Field(function () { return [String]; }, { nullable: true }),
        __metadata("design:type", Array)
    ], KitchenSinkCreateInput.prototype, "arrayOfStrings", void 0);
    __decorate([
        type_graphql_1.Field(function () { return [type_graphql_1.Int]; }, { nullable: true }),
        __metadata("design:type", Array)
    ], KitchenSinkCreateInput.prototype, "arrayOfInts", void 0);
    KitchenSinkCreateInput = __decorate([
        type_graphql_1.InputType()
    ], KitchenSinkCreateInput);
    return KitchenSinkCreateInput;
}());
exports.KitchenSinkCreateInput = KitchenSinkCreateInput;
var KitchenSinkUpdateInput = /** @class */ (function () {
    function KitchenSinkUpdateInput() {
    }
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkUpdateInput.prototype, "stringField", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkUpdateInput.prototype, "nullableStringField", void 0);
    __decorate([
        type_graphql_1.Field(function () { return type_graphql_2.GraphQLISODateTime; }, { nullable: true }),
        __metadata("design:type", Object)
    ], KitchenSinkUpdateInput.prototype, "dateField", void 0);
    __decorate([
        type_graphql_1.Field(function () { return graphql_scalars_1.DateResolver; }, { nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkUpdateInput.prototype, "dateOnlyField", void 0);
    __decorate([
        type_graphql_1.Field(function () { return type_graphql_2.GraphQLISODateTime; }, { nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkUpdateInput.prototype, "dateTimeField", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkUpdateInput.prototype, "emailField", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", Number)
    ], KitchenSinkUpdateInput.prototype, "integerField", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", Boolean)
    ], KitchenSinkUpdateInput.prototype, "booleanField", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", Number)
    ], KitchenSinkUpdateInput.prototype, "floatField", void 0);
    __decorate([
        type_graphql_1.Field(function () { return GraphQLJSONObject; }, { nullable: true }),
        __metadata("design:type", Object)
    ], KitchenSinkUpdateInput.prototype, "jsonField", void 0);
    __decorate([
        type_graphql_1.Field(function () { return graphql_1.GraphQLID; }, { nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkUpdateInput.prototype, "idField", void 0);
    __decorate([
        type_graphql_1.Field(function () { return kitchen_sink_model_1.StringEnum; }, { nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkUpdateInput.prototype, "stringEnumField", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkUpdateInput.prototype, "numericField", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkUpdateInput.prototype, "numericFieldCustomPrecisionScale", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkUpdateInput.prototype, "noFilterField", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkUpdateInput.prototype, "noSortField", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkUpdateInput.prototype, "noFilterOrSortField", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkUpdateInput.prototype, "stringFieldFilterEqContains", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", Number)
    ], KitchenSinkUpdateInput.prototype, "intFieldFilterLteGte", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkUpdateInput.prototype, "characterField", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkUpdateInput.prototype, "customTextFieldNoSortOrFilter", void 0);
    __decorate([
        type_graphql_1.Field(function () { return [String]; }, { nullable: true }),
        __metadata("design:type", Array)
    ], KitchenSinkUpdateInput.prototype, "customFieldArrayColumn", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkUpdateInput.prototype, "writeonlyField", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkUpdateInput.prototype, "apiOnlyField", void 0);
    __decorate([
        type_graphql_1.Field(function () { return [String]; }, { nullable: true }),
        __metadata("design:type", Array)
    ], KitchenSinkUpdateInput.prototype, "arrayOfStrings", void 0);
    __decorate([
        type_graphql_1.Field(function () { return [type_graphql_1.Int]; }, { nullable: true }),
        __metadata("design:type", Array)
    ], KitchenSinkUpdateInput.prototype, "arrayOfInts", void 0);
    KitchenSinkUpdateInput = __decorate([
        type_graphql_1.InputType()
    ], KitchenSinkUpdateInput);
    return KitchenSinkUpdateInput;
}());
exports.KitchenSinkUpdateInput = KitchenSinkUpdateInput;
var KitchenSinkWhereArgs = /** @class */ (function (_super) {
    __extends(KitchenSinkWhereArgs, _super);
    function KitchenSinkWhereArgs() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    __decorate([
        type_graphql_1.Field(function () { return KitchenSinkWhereInput; }, { nullable: true }),
        __metadata("design:type", KitchenSinkWhereInput)
    ], KitchenSinkWhereArgs.prototype, "where", void 0);
    __decorate([
        type_graphql_1.Field(function () { return KitchenSinkOrderByEnum; }, { nullable: true }),
        __metadata("design:type", String)
    ], KitchenSinkWhereArgs.prototype, "orderBy", void 0);
    KitchenSinkWhereArgs = __decorate([
        type_graphql_1.ArgsType()
    ], KitchenSinkWhereArgs);
    return KitchenSinkWhereArgs;
}(__1.PaginationArgs));
exports.KitchenSinkWhereArgs = KitchenSinkWhereArgs;
var KitchenSinkCreateManyArgs = /** @class */ (function () {
    function KitchenSinkCreateManyArgs() {
    }
    __decorate([
        type_graphql_1.Field(function () { return [KitchenSinkCreateInput]; }),
        __metadata("design:type", Array)
    ], KitchenSinkCreateManyArgs.prototype, "data", void 0);
    KitchenSinkCreateManyArgs = __decorate([
        type_graphql_1.ArgsType()
    ], KitchenSinkCreateManyArgs);
    return KitchenSinkCreateManyArgs;
}());
exports.KitchenSinkCreateManyArgs = KitchenSinkCreateManyArgs;
var KitchenSinkUpdateArgs = /** @class */ (function () {
    function KitchenSinkUpdateArgs() {
    }
    __decorate([
        type_graphql_1.Field(),
        __metadata("design:type", KitchenSinkUpdateInput)
    ], KitchenSinkUpdateArgs.prototype, "data", void 0);
    __decorate([
        type_graphql_1.Field(),
        __metadata("design:type", KitchenSinkWhereUniqueInput)
    ], KitchenSinkUpdateArgs.prototype, "where", void 0);
    KitchenSinkUpdateArgs = __decorate([
        type_graphql_1.ArgsType()
    ], KitchenSinkUpdateArgs);
    return KitchenSinkUpdateArgs;
}());
exports.KitchenSinkUpdateArgs = KitchenSinkUpdateArgs;
var DishOrderByEnum;
(function (DishOrderByEnum) {
    DishOrderByEnum["createdAt_ASC"] = "createdAt_ASC";
    DishOrderByEnum["createdAt_DESC"] = "createdAt_DESC";
    DishOrderByEnum["updatedAt_ASC"] = "updatedAt_ASC";
    DishOrderByEnum["updatedAt_DESC"] = "updatedAt_DESC";
    DishOrderByEnum["deletedAt_ASC"] = "deletedAt_ASC";
    DishOrderByEnum["deletedAt_DESC"] = "deletedAt_DESC";
    DishOrderByEnum["name_ASC"] = "name_ASC";
    DishOrderByEnum["name_DESC"] = "name_DESC";
    DishOrderByEnum["stringEnumField_ASC"] = "stringEnumField_ASC";
    DishOrderByEnum["stringEnumField_DESC"] = "stringEnumField_DESC";
    DishOrderByEnum["kitchenSinkId_ASC"] = "kitchenSinkId_ASC";
    DishOrderByEnum["kitchenSinkId_DESC"] = "kitchenSinkId_DESC";
})(DishOrderByEnum = exports.DishOrderByEnum || (exports.DishOrderByEnum = {}));
type_graphql_2.registerEnumType(DishOrderByEnum, {
    name: "DishOrderByInput"
});
var DishWhereInput = /** @class */ (function () {
    function DishWhereInput() {
    }
    __decorate([
        type_graphql_1.Field(function () { return graphql_1.GraphQLID; }, { nullable: true }),
        __metadata("design:type", String)
    ], DishWhereInput.prototype, "id_eq", void 0);
    __decorate([
        type_graphql_1.Field(function () { return [graphql_1.GraphQLID]; }, { nullable: true }),
        __metadata("design:type", Array)
    ], DishWhereInput.prototype, "id_in", void 0);
    __decorate([
        type_graphql_1.Field(function () { return type_graphql_2.GraphQLISODateTime; }, { nullable: true }),
        __metadata("design:type", Object)
    ], DishWhereInput.prototype, "createdAt_eq", void 0);
    __decorate([
        type_graphql_1.Field(function () { return type_graphql_2.GraphQLISODateTime; }, { nullable: true }),
        __metadata("design:type", Object)
    ], DishWhereInput.prototype, "createdAt_lt", void 0);
    __decorate([
        type_graphql_1.Field(function () { return type_graphql_2.GraphQLISODateTime; }, { nullable: true }),
        __metadata("design:type", Object)
    ], DishWhereInput.prototype, "createdAt_lte", void 0);
    __decorate([
        type_graphql_1.Field(function () { return type_graphql_2.GraphQLISODateTime; }, { nullable: true }),
        __metadata("design:type", Object)
    ], DishWhereInput.prototype, "createdAt_gt", void 0);
    __decorate([
        type_graphql_1.Field(function () { return type_graphql_2.GraphQLISODateTime; }, { nullable: true }),
        __metadata("design:type", Object)
    ], DishWhereInput.prototype, "createdAt_gte", void 0);
    __decorate([
        type_graphql_1.Field(function () { return graphql_1.GraphQLID; }, { nullable: true }),
        __metadata("design:type", String)
    ], DishWhereInput.prototype, "createdById_eq", void 0);
    __decorate([
        type_graphql_1.Field(function () { return [graphql_1.GraphQLID]; }, { nullable: true }),
        __metadata("design:type", Array)
    ], DishWhereInput.prototype, "createdById_in", void 0);
    __decorate([
        type_graphql_1.Field(function () { return type_graphql_2.GraphQLISODateTime; }, { nullable: true }),
        __metadata("design:type", Object)
    ], DishWhereInput.prototype, "updatedAt_eq", void 0);
    __decorate([
        type_graphql_1.Field(function () { return type_graphql_2.GraphQLISODateTime; }, { nullable: true }),
        __metadata("design:type", Object)
    ], DishWhereInput.prototype, "updatedAt_lt", void 0);
    __decorate([
        type_graphql_1.Field(function () { return type_graphql_2.GraphQLISODateTime; }, { nullable: true }),
        __metadata("design:type", Object)
    ], DishWhereInput.prototype, "updatedAt_lte", void 0);
    __decorate([
        type_graphql_1.Field(function () { return type_graphql_2.GraphQLISODateTime; }, { nullable: true }),
        __metadata("design:type", Object)
    ], DishWhereInput.prototype, "updatedAt_gt", void 0);
    __decorate([
        type_graphql_1.Field(function () { return type_graphql_2.GraphQLISODateTime; }, { nullable: true }),
        __metadata("design:type", Object)
    ], DishWhereInput.prototype, "updatedAt_gte", void 0);
    __decorate([
        type_graphql_1.Field(function () { return graphql_1.GraphQLID; }, { nullable: true }),
        __metadata("design:type", String)
    ], DishWhereInput.prototype, "updatedById_eq", void 0);
    __decorate([
        type_graphql_1.Field(function () { return [graphql_1.GraphQLID]; }, { nullable: true }),
        __metadata("design:type", Array)
    ], DishWhereInput.prototype, "updatedById_in", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", Boolean)
    ], DishWhereInput.prototype, "deletedAt_all", void 0);
    __decorate([
        type_graphql_1.Field(function () { return type_graphql_2.GraphQLISODateTime; }, { nullable: true }),
        __metadata("design:type", Object)
    ], DishWhereInput.prototype, "deletedAt_eq", void 0);
    __decorate([
        type_graphql_1.Field(function () { return type_graphql_2.GraphQLISODateTime; }, { nullable: true }),
        __metadata("design:type", Object)
    ], DishWhereInput.prototype, "deletedAt_lt", void 0);
    __decorate([
        type_graphql_1.Field(function () { return type_graphql_2.GraphQLISODateTime; }, { nullable: true }),
        __metadata("design:type", Object)
    ], DishWhereInput.prototype, "deletedAt_lte", void 0);
    __decorate([
        type_graphql_1.Field(function () { return type_graphql_2.GraphQLISODateTime; }, { nullable: true }),
        __metadata("design:type", Object)
    ], DishWhereInput.prototype, "deletedAt_gt", void 0);
    __decorate([
        type_graphql_1.Field(function () { return type_graphql_2.GraphQLISODateTime; }, { nullable: true }),
        __metadata("design:type", Object)
    ], DishWhereInput.prototype, "deletedAt_gte", void 0);
    __decorate([
        type_graphql_1.Field(function () { return graphql_1.GraphQLID; }, { nullable: true }),
        __metadata("design:type", String)
    ], DishWhereInput.prototype, "deletedById_eq", void 0);
    __decorate([
        type_graphql_1.Field(function () { return [graphql_1.GraphQLID]; }, { nullable: true }),
        __metadata("design:type", Array)
    ], DishWhereInput.prototype, "deletedById_in", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], DishWhereInput.prototype, "name_eq", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], DishWhereInput.prototype, "name_contains", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], DishWhereInput.prototype, "name_startsWith", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], DishWhereInput.prototype, "name_endsWith", void 0);
    __decorate([
        type_graphql_1.Field(function () { return [String]; }, { nullable: true }),
        __metadata("design:type", Array)
    ], DishWhereInput.prototype, "name_in", void 0);
    __decorate([
        type_graphql_1.Field(function () { return kitchen_sink_model_1.StringEnum; }, { nullable: true }),
        __metadata("design:type", String)
    ], DishWhereInput.prototype, "stringEnumField_eq", void 0);
    __decorate([
        type_graphql_1.Field(function () { return [kitchen_sink_model_1.StringEnum]; }, { nullable: true }),
        __metadata("design:type", Array)
    ], DishWhereInput.prototype, "stringEnumField_in", void 0);
    __decorate([
        type_graphql_1.Field(function () { return graphql_1.GraphQLID; }, { nullable: true }),
        __metadata("design:type", String)
    ], DishWhereInput.prototype, "kitchenSinkId_eq", void 0);
    __decorate([
        type_graphql_1.Field(function () { return [graphql_1.GraphQLID]; }, { nullable: true }),
        __metadata("design:type", Array)
    ], DishWhereInput.prototype, "kitchenSinkId_in", void 0);
    DishWhereInput = __decorate([
        type_graphql_1.InputType()
    ], DishWhereInput);
    return DishWhereInput;
}());
exports.DishWhereInput = DishWhereInput;
var DishWhereUniqueInput = /** @class */ (function () {
    function DishWhereUniqueInput() {
    }
    __decorate([
        type_graphql_1.Field(function () { return graphql_1.GraphQLID; }),
        __metadata("design:type", String)
    ], DishWhereUniqueInput.prototype, "id", void 0);
    DishWhereUniqueInput = __decorate([
        type_graphql_1.InputType()
    ], DishWhereUniqueInput);
    return DishWhereUniqueInput;
}());
exports.DishWhereUniqueInput = DishWhereUniqueInput;
var DishCreateInput = /** @class */ (function () {
    function DishCreateInput() {
    }
    __decorate([
        type_graphql_1.Field(),
        __metadata("design:type", String)
    ], DishCreateInput.prototype, "name", void 0);
    __decorate([
        type_graphql_1.Field(function () { return kitchen_sink_model_1.StringEnum; }, { nullable: true }),
        __metadata("design:type", String)
    ], DishCreateInput.prototype, "stringEnumField", void 0);
    __decorate([
        type_graphql_1.Field(function () { return graphql_1.GraphQLID; }),
        __metadata("design:type", String)
    ], DishCreateInput.prototype, "kitchenSinkId", void 0);
    DishCreateInput = __decorate([
        type_graphql_1.InputType()
    ], DishCreateInput);
    return DishCreateInput;
}());
exports.DishCreateInput = DishCreateInput;
var DishUpdateInput = /** @class */ (function () {
    function DishUpdateInput() {
    }
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], DishUpdateInput.prototype, "name", void 0);
    __decorate([
        type_graphql_1.Field(function () { return kitchen_sink_model_1.StringEnum; }, { nullable: true }),
        __metadata("design:type", String)
    ], DishUpdateInput.prototype, "stringEnumField", void 0);
    __decorate([
        type_graphql_1.Field(function () { return graphql_1.GraphQLID; }, { nullable: true }),
        __metadata("design:type", String)
    ], DishUpdateInput.prototype, "kitchenSinkId", void 0);
    DishUpdateInput = __decorate([
        type_graphql_1.InputType()
    ], DishUpdateInput);
    return DishUpdateInput;
}());
exports.DishUpdateInput = DishUpdateInput;
var DishWhereArgs = /** @class */ (function (_super) {
    __extends(DishWhereArgs, _super);
    function DishWhereArgs() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    __decorate([
        type_graphql_1.Field(function () { return DishWhereInput; }, { nullable: true }),
        __metadata("design:type", DishWhereInput)
    ], DishWhereArgs.prototype, "where", void 0);
    __decorate([
        type_graphql_1.Field(function () { return DishOrderByEnum; }, { nullable: true }),
        __metadata("design:type", String)
    ], DishWhereArgs.prototype, "orderBy", void 0);
    DishWhereArgs = __decorate([
        type_graphql_1.ArgsType()
    ], DishWhereArgs);
    return DishWhereArgs;
}(__1.PaginationArgs));
exports.DishWhereArgs = DishWhereArgs;
var DishCreateManyArgs = /** @class */ (function () {
    function DishCreateManyArgs() {
    }
    __decorate([
        type_graphql_1.Field(function () { return [DishCreateInput]; }),
        __metadata("design:type", Array)
    ], DishCreateManyArgs.prototype, "data", void 0);
    DishCreateManyArgs = __decorate([
        type_graphql_1.ArgsType()
    ], DishCreateManyArgs);
    return DishCreateManyArgs;
}());
exports.DishCreateManyArgs = DishCreateManyArgs;
var DishUpdateArgs = /** @class */ (function () {
    function DishUpdateArgs() {
    }
    __decorate([
        type_graphql_1.Field(),
        __metadata("design:type", DishUpdateInput)
    ], DishUpdateArgs.prototype, "data", void 0);
    __decorate([
        type_graphql_1.Field(),
        __metadata("design:type", DishWhereUniqueInput)
    ], DishUpdateArgs.prototype, "where", void 0);
    DishUpdateArgs = __decorate([
        type_graphql_1.ArgsType()
    ], DishUpdateArgs);
    return DishUpdateArgs;
}());
exports.DishUpdateArgs = DishUpdateArgs;
