// warthog/dist/test/server-vars.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.clearConfig = exports.getStandardEnvironmentVariables = exports.setTestServerEnvironmentVariables = void 0;
function setTestServerEnvironmentVariables(overrides) {
    clearConfig();
    var defaultVars = getStandardEnvironmentVariables();
    Object.keys(defaultVars).forEach(function (key) {
        process.env[key] = defaultVars[key];
    });
    if (!overrides) {
        return;
    }
    Object.keys(overrides).forEach(function (key) {
        process.env[key] = overrides[key];
    });
}
exports.setTestServerEnvironmentVariables = setTestServerEnvironmentVariables;
function getStandardEnvironmentVariables() {
    return {
        WARTHOG_APP_HOST: 'localhost',
        WARTHOG_APP_PORT: '4000',
        WARTHOG_APP_PROTOCOL: 'http',
        WARTHOG_AUTO_GENERATE_FILES: 'false',
        WARTHOG_AUTO_OPEN_PLAYGROUND: 'false',
        WARTHOG_DB_DATABASE: 'warthog-test',
        WARTHOG_DB_ENTITIES: 'src/test/modules/**/*.model.ts',
        WARTHOG_DB_HOST: 'localhost',
        WARTHOG_DB_LOGGING: 'none',
        WARTHOG_DB_MIGRATIONS_DIR: './tmp/test/migrations',
        WARTHOG_DB_OVERRIDE: 'true',
        WARTHOG_DB_USERNAME: 'postgres',
        WARTHOG_DB_PASSWORD: '',
        WARTHOG_DB_SYNCHRONIZE: 'true',
        WARTHOG_GENERATED_FOLDER: './src/test/generated',
        WARTHOG_RESOLVERS_PATH: './src/test/modules/**/*.resolver.ts',
        WARTHOG_MODULE_IMPORT_PATH: '../../',
        WARTHOG_VALIDATE_RESOLVERS: 'false'
    };
}
exports.getStandardEnvironmentVariables = getStandardEnvironmentVariables;
function clearConfig() {
    var WARTHOG_PREFIX = 'WARTHOG_';
    var TYPEORM_PREFIX = 'TYPEORM_';
    Object.keys(process.env).forEach(function (key) {
        if (key.startsWith(WARTHOG_PREFIX) || key.startsWith(TYPEORM_PREFIX)) {
            delete process.env[key];
        }
    });
}
exports.clearConfig = clearConfig;
