// warthog/dist/test/examples/create-id-model/generated/classes.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CreateIDTestUpdateArgs = exports.CreateIDTestCreateManyArgs = exports.CreateIDTestWhereArgs = exports.CreateIDTestUpdateInput = exports.CreateIDTestCreateInput = exports.CreateIDTestWhereUniqueInput = exports.CreateIDTestWhereInput = exports.CreateIDTestOrderByEnum = void 0;
// This file has been auto-generated by Warthog.  Do not update directly as it
// will be re-written.  If you need to change this file, update models or add
// new TypeGraphQL objects
var type_graphql_1 = require("type-graphql");
var type_graphql_2 = require("type-graphql");
var __1 = require("../../../../");
var CreateIDTestOrderByEnum;
(function (CreateIDTestOrderByEnum) {
    CreateIDTestOrderByEnum["createdAt_ASC"] = "createdAt_ASC";
    CreateIDTestOrderByEnum["createdAt_DESC"] = "createdAt_DESC";
    CreateIDTestOrderByEnum["updatedAt_ASC"] = "updatedAt_ASC";
    CreateIDTestOrderByEnum["updatedAt_DESC"] = "updatedAt_DESC";
    CreateIDTestOrderByEnum["deletedAt_ASC"] = "deletedAt_ASC";
    CreateIDTestOrderByEnum["deletedAt_DESC"] = "deletedAt_DESC";
    CreateIDTestOrderByEnum["stringField_ASC"] = "stringField_ASC";
    CreateIDTestOrderByEnum["stringField_DESC"] = "stringField_DESC";
})(CreateIDTestOrderByEnum = exports.CreateIDTestOrderByEnum || (exports.CreateIDTestOrderByEnum = {}));
type_graphql_2.registerEnumType(CreateIDTestOrderByEnum, {
    name: "CreateIDTestOrderByInput"
});
var CreateIDTestWhereInput = /** @class */ (function (_super) {
    __extends(CreateIDTestWhereInput, _super);
    function CreateIDTestWhereInput() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], CreateIDTestWhereInput.prototype, "stringField_eq", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], CreateIDTestWhereInput.prototype, "stringField_contains", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], CreateIDTestWhereInput.prototype, "stringField_startsWith", void 0);
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], CreateIDTestWhereInput.prototype, "stringField_endsWith", void 0);
    __decorate([
        type_graphql_1.Field(function () { return [String]; }, { nullable: true }),
        __metadata("design:type", Array)
    ], CreateIDTestWhereInput.prototype, "stringField_in", void 0);
    CreateIDTestWhereInput = __decorate([
        type_graphql_1.InputType()
    ], CreateIDTestWhereInput);
    return CreateIDTestWhereInput;
}(__1.BaseWhereInput));
exports.CreateIDTestWhereInput = CreateIDTestWhereInput;
var CreateIDTestWhereUniqueInput = /** @class */ (function () {
    function CreateIDTestWhereUniqueInput() {
    }
    __decorate([
        type_graphql_1.Field(function () { return String; }),
        __metadata("design:type", String)
    ], CreateIDTestWhereUniqueInput.prototype, "id", void 0);
    CreateIDTestWhereUniqueInput = __decorate([
        type_graphql_1.InputType()
    ], CreateIDTestWhereUniqueInput);
    return CreateIDTestWhereUniqueInput;
}());
exports.CreateIDTestWhereUniqueInput = CreateIDTestWhereUniqueInput;
var CreateIDTestCreateInput = /** @class */ (function () {
    function CreateIDTestCreateInput() {
    }
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], CreateIDTestCreateInput.prototype, "id", void 0);
    __decorate([
        type_graphql_1.Field(),
        __metadata("design:type", String)
    ], CreateIDTestCreateInput.prototype, "stringField", void 0);
    CreateIDTestCreateInput = __decorate([
        type_graphql_1.InputType()
    ], CreateIDTestCreateInput);
    return CreateIDTestCreateInput;
}());
exports.CreateIDTestCreateInput = CreateIDTestCreateInput;
var CreateIDTestUpdateInput = /** @class */ (function () {
    function CreateIDTestUpdateInput() {
    }
    __decorate([
        type_graphql_1.Field({ nullable: true }),
        __metadata("design:type", String)
    ], CreateIDTestUpdateInput.prototype, "stringField", void 0);
    CreateIDTestUpdateInput = __decorate([
        type_graphql_1.InputType()
    ], CreateIDTestUpdateInput);
    return CreateIDTestUpdateInput;
}());
exports.CreateIDTestUpdateInput = CreateIDTestUpdateInput;
var CreateIDTestWhereArgs = /** @class */ (function (_super) {
    __extends(CreateIDTestWhereArgs, _super);
    function CreateIDTestWhereArgs() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    __decorate([
        type_graphql_1.Field(function () { return CreateIDTestWhereInput; }, { nullable: true }),
        __metadata("design:type", CreateIDTestWhereInput)
    ], CreateIDTestWhereArgs.prototype, "where", void 0);
    __decorate([
        type_graphql_1.Field(function () { return CreateIDTestOrderByEnum; }, { nullable: true }),
        __metadata("design:type", String)
    ], CreateIDTestWhereArgs.prototype, "orderBy", void 0);
    CreateIDTestWhereArgs = __decorate([
        type_graphql_1.ArgsType()
    ], CreateIDTestWhereArgs);
    return CreateIDTestWhereArgs;
}(__1.PaginationArgs));
exports.CreateIDTestWhereArgs = CreateIDTestWhereArgs;
var CreateIDTestCreateManyArgs = /** @class */ (function () {
    function CreateIDTestCreateManyArgs() {
    }
    __decorate([
        type_graphql_1.Field(function () { return [CreateIDTestCreateInput]; }),
        __metadata("design:type", Array)
    ], CreateIDTestCreateManyArgs.prototype, "data", void 0);
    CreateIDTestCreateManyArgs = __decorate([
        type_graphql_1.ArgsType()
    ], CreateIDTestCreateManyArgs);
    return CreateIDTestCreateManyArgs;
}());
exports.CreateIDTestCreateManyArgs = CreateIDTestCreateManyArgs;
var CreateIDTestUpdateArgs = /** @class */ (function () {
    function CreateIDTestUpdateArgs() {
    }
    __decorate([
        type_graphql_1.Field(),
        __metadata("design:type", CreateIDTestUpdateInput)
    ], CreateIDTestUpdateArgs.prototype, "data", void 0);
    __decorate([
        type_graphql_1.Field(),
        __metadata("design:type", CreateIDTestWhereUniqueInput)
    ], CreateIDTestUpdateArgs.prototype, "where", void 0);
    CreateIDTestUpdateArgs = __decorate([
        type_graphql_1.ArgsType()
    ], CreateIDTestUpdateArgs);
    return CreateIDTestUpdateArgs;
}());
exports.CreateIDTestUpdateArgs = CreateIDTestUpdateArgs;
